katdal/categorical.py

```diff
@@ -16,124 +16,135 @@
 
 """Container for categorical (i.e. non-numerical) sensor data and related tools."""
 
+import collections
+import cPickle as pickle
+
 import numpy as np
 
 
-def unique(ar, return_index=False, return_inverse=False):
-    """Find the unique elements of an array.
+class HashableValueWrapper(object):
+    """Wrapper that ensures objects are hashable and comparable by value.
+
+    This wrapper class has two main benefits:
 
-    Returns the sorted unique elements of an array. There are two optional
-    outputs in addition to the unique elements: the indices of the input array
-    that give the unique values, and the indices of the unique array that
-    reconstruct the input array.
+      - It prevents sensor values that are NumPy ndarrays themselves (or
+        array-like objects such as tuples and lists) from dissolving and
+        losing their identity when they are assembled into an array.
+
+      - It ensures that all sensor values become hashable by value, which eases
+        comparison (e.g. avoiding the array-valued booleans resulting from
+        ndarray comparisons without resorting to np.array_equal) and also
+        allows sensor values to be looked up in mappings to simplify indexing.
 
     Parameters
     ----------
-    ar : array_like
-        Input array. This will be flattened if it is not already 1-D.
-    return_index : bool, optional
-        If True, also return the indices of `ar` that result in the unique
-        array.
-    return_inverse : bool, optional
-        If True, also return the indices of the unique array that can be used
-        to reconstruct `ar`.
+    hashable_value : object that is hashable
+        The wrapped version of a sensor value - if not available, use static
+        method :meth:`wrap` instead (current implementation expects a pickle)
 
-    Returns
-    -------
-    unique : ndarray
-        The sorted unique values.
-    unique_indices : ndarray, optional
-        The indices of the first occurrences of the unique values in the
-        (flattened) original array. Only provided if `return_index` is True.
-    unique_inverse : ndarray, optional
-        The indices to reconstruct the (flattened) original array from the
-        unique array. Only provided if `return_inverse` is True.
+    Note
+    ----
+    The current implementation uses pickle to make generic objects hashable,
+    even though it is not recommended (see http://bit.ly/2ruNkH2). The main
```

Comment:
Rather put in the original URL, so that this won't break if bit.ly goes down or goes out of business.